### DISCUSSION TITLE
Error with custom upload provider

### DISCUSSION THREAD

**Discussion by chrissnoek** [strapi/strapi#6058] [Legacy Topics]
**Describe the bug**
I created a custom upload provider, but i get a very vague error saying 
`[2020-05-04T19:19:05.455Z] error ValidationError: file validation failed: url: Path `url` is required.
    at new ValidationError (/Users/chrissnoek/Documents/#NPM PROJECTS/spotshareWebsite/strapi/spotshare/node_modules/mongoose/lib/error/validation.js:31:11)`

**Steps to reproduce the behavior**
Use this as a custom upload provider, and upload an image through the admin
`'use strict';

/* eslint-disable no-unused-vars */
// Public node modules.
const AWS = require("aws-sdk");
const sharp = require("sharp");
const path = require("path");

module.exports = {
  init(providerOptions) {

    AWS.config.update({
      apiVersion: '2006-03-01',
      ...providerOptions
    });

    const S3 = new AWS.S3();

    return {
      async upload(file) {

        console.log(file);
        const generatedFiles = await generateFiles(file);

        generatedFiles.forEach(image => {
          //console.log(image);

          const { buffer, mime, suffix } = image;

          return new Promise((resolve, reject) => {
            const path = file.path ? `${file.path}/` : '';
            //const objectKey = `${filePath}${path.parse(file.originalname).name}${suffix}`;
            const objectKey = `${file.name}${suffix}${file.ext}`;

            S3.upload(
              {
                Key: `${path}${objectKey}`,
                Body: Buffer.from(file.buffer, 'binary'),
                ACL: 'public-read',
                ContentType: file.mime,
                Bucket: "spotsharenl"
              },
              (err, data) => {
                if (err) {
                  return reject(err);
                }

                // set the bucket file url
                file.url = data.Location;

                resolve();
              }
            );

          });
        });

      },
      delete(file) {

        let generatedSuffixes;

        switch (file.ext.toLowerCase()) {
          case '.png':
            generatedSuffixes = [`-original.png`, `-watermark.png`, `-thumbnail.png`, `-small.png`];
            break;
          case '.jpg':
          case '.jpeg':
            generatedSuffixes = [`-original.jpg`, `-watermark.jpg`, `-thumbnail.jpg`, `-small.jpg`];
            break;
        }

        generatedSuffixes.forEach(suffix => {
          return new Promise((resolve, reject) => {
            const path = file.path ? `${file.path}/` : '';
            const objectKey = `${file.name}${suffix}${file.ext}`;

            S3.deleteObject({
              Key: `${path}${objectKey}`,
            }, (err, data) => {
              if (err) {
                return reject(err);
              }
              resolve();
            }
            );
          });
        });

      },
    };
  },
};

const generateFiles = async (file) => {
  console.log('in generatedFiles');
  const buffer = new Buffer.from(file.buffer, 'binary');

  console.log('made buffer');

  let images = [];

  console.log(file.ext.toLowerCase());

  switch (file.ext.toLowerCase()) {
    case '.png':

      //original
      images.push(
        await sharp(buffer)
          .png().toBuffer()
          .then(data => ({
            buffer: data,
            mime: file.mime,
            ext: file.ext,
            suffix: `-original`
          }))
      );

      //thumbnail
      images.push(
        await sharp(buffer)
          .resize(800, 400)
          .png().toBuffer()
          .then(data => ({
            buffer: data,
            mime: file.mime,
            ext: file.ext,
            suffix: `-thumbnail`
          }))
      );

      //watermark
      images.push(
        await sharp(buffer)
          .resize(1500)
          .composite([
            {
              input: "./overlay.png",
              gravity: "southwest"
            }
          ]).png().toBuffer()
          .then(data => ({
            buffer: data,
            mime: file.mime,
            ext: file.ext,
            suffix: `-watermark`
          }))
      );

      //small
      images.push(
        await sharp(buffer)
          .resize(500)
          .png().toBuffer()
          .then(data => ({
            buffer: data,
            mime: file.mime,
            ext: file.ext,
            suffix: `-small`
          }))
      );

      break;
    case '.jpg':
    case '.jpeg':

      //original
      images.push(
        await sharp(buffer)
          .jpeg().toBuffer()
          .then(data => ({
            buffer: data,
            mime: file.mime,
            ext: file.ext,
            suffix: `-original`
          }))
      );

      //thumbnail
      images.push(
        await sharp(buffer)
          .resize(800, 400)
          .jpeg().toBuffer()
          .then(data => ({
            buffer: data,
            mime: file.mime,
            ext: file.ext,
            suffix: `-thumbnail`
          }))
      );

      //watermark
      images.push(
        await sharp(buffer)
          .resize(1500)
          .composite([
            {
              input: "./overlay.png",
              gravity: "southwest"
            }
          ]).jpeg().toBuffer()
          .then(data => ({
            buffer: data,
            mime: file.mime,
            ext: file.ext,
            suffix: `-watermark`
          }))
      );

      //watermark
      images.push(
        await sharp(buffer)
          .resize(500)
          .jpeg().toBuffer()
          .then(data => ({
            buffer: data,
            mime: file.mime,
            ext: file.ext,
            suffix: `-small`
          }))
      );

      break;
  }
  //console.log(images);
  return Promise.all(images);
}
`

**Expected behavior**
Upload the images in various sizes.

**System**

- Node.js version: v12.14.0
- NPM version: 6.13.4
- Strapi version: v3.0.0-beta.20.1
- Database: MongoDB
- Operating system: MacOS

What am i missing? Please help!

**Comment by derrickmehaffy** [comment]
Hello,

I see you are wanting to ask a question that is not really a bug report or a feature request, questions should be directed to the [discussions](https://github.com/strapi/strapi/discussions)

I have converted this to a discussion for you, we try to keep Github clean and only for bug reports and feature requests.

**Comment by chrissnoek** [comment]
Ok, i found my problem. I was generating new files with the function, but my original file variable was still given at the S3 upload, instead of the generated image buffer..
